File: src/0375-Guess-Number-Higher-or-Lower-II/0375.py
class Solution:
    def getMoneyAmount(self, n):
        dp = [[0] * (n+1) for _ in range(n+1)]
        for L in range(n-1, 0, -1):
            for R in range(L+1, n+1):
                dp[L][R] = min(i + max(dp[L][i-1], dp[i+1][R]) for i in range(L, R))
        return dp[1][n]


if __name__ == "__main__":
    print Solution().getMoneyAmount(10)

File: src/0375-Guess-Number-Higher-or-Lower-II/0375.scala
object Solution{
  def getMontyAmount(n:Int):Int = {
    val dp = Array.ofDim[Int](n+1,n+1)
    for(l <- (n-1).to(1).by(-1)){
      for(r <- (l+1).to(n).by(1)){
        var tmp = List[Int]()
        for(i <- l.to(r-1).by(1)) {
          tmp ::= (i + (dp(l)(i-1)).max(dp(i+1)(r)))
        }
        dp(l)(r) = tmp.min
      }
    }
    dp(1)(n)
  }

  def main(args:Array[String]): Unit ={
    val n = 10
    println(getMontyAmount(n))
  }
}